tests/cpp/test_task_graph.cpp

```diff
@@ -0,0 +1,269 @@
+  std::vector<TaskGraph::TaskId> tasks;
+  tasks.reserve(steps.size());
+  for (const TaskGraph::Step& step : steps) {
+    tasks.push_back(step.task);
+  }
+  return tasks;
+}
+
+TEST_F(TaskGraphTest, Basic) {
+  //   0   1
+  //   |\ /
+  //   | 2
+  //   |/
+  //   3
+  Tasks tasks{{{0, 1}, {2}}, {{0, 2}, {3}}};
+  std::vector<TaskGraph::Data> data = inferData(tasks);
+  auto graph = TaskGraph(tasks, data);
+
+  const TaskGraph::SortResult result = graph.findOptimalOrder();
+
+  ASSERT_EQ(result.steps.size(), tasks.size());
+  std::vector<TaskGraph::TaskId> expected{0, 1};
+  EXPECT_EQ(getTasks(result), expected);
+  EXPECT_EQ(result.steps.back().high_water_mark, 4);
+}
+
+// This example includes two segments, each of which aliases the other
+TEST_F(TaskGraphTest, ImpossibleAlias) {
+  //   0   1
+  //   |\ /|
+  //   | X |
+  //   |/ \|
+  //   2   3
+  // Two tasks, each takes the same two inputs
+  Tasks tasks{{{0, 1}, {2}}, {{0, 1}, {3}}};
+  std::vector<TaskGraph::Data> data = inferData(tasks);
+  // Each of the segment outputs aliases a different input
+  data[2].aliases_input = 0;
+  data[3].aliases_input = 1;
+  // This graph can't be ordered without breaking the aliasing constraint
+  auto graph = TaskGraph(tasks, data);
+
+  EXPECT_THAT(
+      [&graph]() { graph.findOptimalOrder(); },
+      ::testing::ThrowsMessage<nvfuser::nvfError>(::testing::HasSubstr(
+          "Ran out of ready tasks before completing ordering")));
+}
+
+TEST_F(TaskGraphTest, SelfEdge) {
+  Tasks tasks{{{0}, {0}}};
+  std::vector<TaskGraph::Data> data = inferData(tasks);
+  // This graph can't be ordered because it contains an edge from a Data node
+  // back to itself. A task can't be both producer and consumer to a Data.
+  auto graph = TaskGraph(tasks, data);
+
+  EXPECT_THAT(
+      [&graph]() { graph.findOptimalOrder(); },
+      ::testing::ThrowsMessage<nvfuser::nvfError>(::testing::HasSubstr(
+          "Ran out of ready tasks before completing ordering")));
+}
+
+TEST_F(TaskGraphTest, TwoCycle) {
+  Tasks tasks{{{0}, {1}}, {{1}, {0}}};
+  std::vector<TaskGraph::Data> data = inferData(tasks);
+  // This graph can't be ordered because it contains a cycle
+  auto graph = TaskGraph(tasks, data);
+
+  EXPECT_THAT(
+      [&graph]() { graph.findOptimalOrder(); },
+      ::testing::ThrowsMessage<nvfuser::nvfError>(::testing::HasSubstr(
+          "Ran out of ready tasks before completing ordering")));
+}
+
+TEST_F(TaskGraphTest, ThreeCycle) {
+  Tasks tasks{{{0}, {1}}, {{1}, {2}}, {{2}, {0}}};
+  std::vector<TaskGraph::Data> data = inferData(tasks);
+  // This graph can't be ordered because it contains a cycle
+  auto graph = TaskGraph(tasks, data);
+
+  EXPECT_THAT(
+      [&graph]() { graph.findOptimalOrder(); },
+      ::testing::ThrowsMessage<nvfuser::nvfError>(::testing::HasSubstr(
+          "Ran out of ready tasks before completing ordering")));
+}
+
+TEST_F(TaskGraphTest, FreeableIntermediate) {
+  //   0
+  //  /|\
```

Comment:
```suggestion
  //  /|\. 
```
**[Suggested by AI]** Summary of the fix
• A single character was added to the end of a comment: the line  
    //  /|\  
  became  
    //  /|\.

• The original comment ended with a back-slash. In C/C++ source this is dangerous, because a back-slash that is the last character on a physical line is interpreted by the pre-processor as a line-continuation marker. The result is that the newline is removed and the “//” comment inadvertently spills over onto the following line, making the compiler ignore real code and producing build errors.

• By appending a harmless period after the back-slash, the back-slash is no longer the final character on the line, so the line-continuation rule is not triggered. The comment is now confined to its own line and compilation succeeds.